care/users/models.py

```diff
@@ -405,7 +406,23 @@ def has_add_user_permission(request):
     def check_username_exists(username):
         return User.objects.get_entire_queryset().filter(username=username).exists()
 
+    @transaction.atomic
     def delete(self, *args, **kwargs):
+        from care.facility.models.asset import UserDefaultAssetLocation
+        from care.facility.models.facility import FacilityUser
+        from care.facility.models.patient_sample import PatientSample
+
+        if FacilityUser.objects.filter(created_by=self).exists():
+            error = f"Cannot delete User {self.username} because they are referenced as `created_by` in FacilityUser records."
+            raise ValidationError(error)
+
+        if UserDefaultAssetLocation.objects.filter(user=self).exists():
+            error = f"Cannot delete User {self} because they are referenced as `user` in UserDefaultAssetLocation records."
+            raise ValidationError(error)
+
+        FacilityUser.objects.filter(user=self).delete()
+        PatientSample.objects.filter(created_by=self).update(created_by=None)
+        PatientSample.objects.filter(last_edited_by=self).update(last_edited_by=None)
```

Comment:
WE can keep these, the whole reason we dont delete users is so that we can show these when the audit comes up.

Reply:
```
PatientSample.objects.filter(created_by=self).update(created_by=None)
PatientSample.objects.filter(last_edited_by=self).update(last_edited_by=None)
```
You mean that we don't need to set user null here?
@vigneshhari